KAFKA-14367; Add `JoinGroup` to the new `GroupCoordinator` interface

dajac opened the pull request:
This patch adds `joinGroup` to the new `GroupCoordinator` interface and updates `KafkaApis` to use it.

For the context, I will do the same for all the other interactions with the current group coordinator. In order to limit the changes, I have chosen to introduce the `GroupCoordinatorAdapter` that translates the new interface to the old one. It is basically a wrapper. This allows keeping the current group coordinator untouched for now and focus on the `KafkaApis` changes. Eventually, we can remove `GroupCoordinatorAdapter`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

dajac committed:
KAFKA-14367; Add `JoinGroup` to the new `GroupCoordinator` interface

Add GroupCoordinator interface

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1646,73 +1649,46 @@ class KafkaApis(val requestChannel: RequestChannel,
-          joinResult.protocolName.getOrElse(GroupCoordinator.NoProtocol)
-
-        val responseBody = new JoinGroupResponse(
-          new JoinGroupResponseData()
```
This translation is now in `GroupCoordinatorAdapter`.

jolshan replied:
Looks like we've also simplified building the response. Looks good.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1646,73 +1649,46 @@ class KafkaApis(val requestChannel: RequestChannel,
-      sendResponseCallback(JoinGroupResult(JoinGroupRequest.UNKNOWN_MEMBER_ID, Errors.GROUP_AUTHORIZATION_FAILED))
+      sendResponse(joinGroupRequest.getErrorResponse(Errors.GROUP_AUTHORIZATION_FAILED.exception))
     } else {
-      val groupInstanceId = Option(joinGroupRequest.data.groupInstanceId)
```
This code is not in `GroupCoordinatorAdaptor` without any changes.

jolshan replied:
Did you mean that it is in the adapter? I see it at line 50

dajac replied:
Yeah, I meant `now` instead of `not`. This code was moved 1:1 there.

dajac commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -94,6 +95,7 @@ class KafkaApis(val requestChannel: RequestChannel,
                 val metadataSupport: MetadataSupport,
                 val replicaManager: ReplicaManager,
                 val groupCoordinator: GroupCoordinator,
+                val newGroupCoordinator: org.apache.kafka.coordinator.group.GroupCoordinator,
```
This is a temporary change. When all the API are migrated to new new interface, I will change `groupCoordinator` to the new interface and remove this one.

jeffkbkim replied:
should we add a comment?

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2528,166 @@ class KafkaApisTest {
   }
 
-  @Test
-  def testJoinGroupProtocolsOrder(): Unit = {
```
This test's logic is now tested in `GroupCoordinatorAdaptorTest`.

jeffkbkim replied:
do you mean "now"?

dajac replied:
Right. Let me fix it.

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2528,166 @@ class KafkaApisTest {
-  }
 
-  @Test
-  def testJoinGroupWhenAnErrorOccurs(): Unit = {
```
I have renames this one to `testJoinGroupProtocolTypeBackwardCompatibility`.

jolshan replied:
Git reallly made the comparison hard for these tests. 😅

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2528,166 @@ class KafkaApisTest {
   }
 
   @Test
-  def testJoinGroupProtocolType(): Unit = {
```
This is now named `testHandleJoinGroupRequest`.

jolshan replied:
I'm not sure I follow here.

dajac replied:
Sorry for the typo. I meant `now` instead of `not`, again.

dajac commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2528,166 @@ class KafkaApisTest {
 
-    createKafkaApis().handleJoinGroupRequest(requestChannelRequest, RequestLocal.withThreadConfinedCaching)
+  @Test
+  def testHandleJoinGroupRequestAuthorizationFailed(): Unit = {
```
This is a new test. This code path was not tested before from `KafkaApisTest`.

hachikuji commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorRequestContext.java:
```diff
@@ -0,0 +1,74 @@
+
+public class GroupCoordinatorRequestContext {
+
+    private final short apiVersion;
```
Perhaps we should use `RequestHeader` like the other request context objects. It's a little weird to see the api version without the corresponding api key.

dajac replied:
Yeah, I agree with you. I went with the minimal to start with.

hachikuji commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorRequestContext.java:
```diff
@@ -0,0 +1,74 @@
+import java.net.InetAddress;
+import java.util.Objects;
+
+public class GroupCoordinatorRequestContext {
```
I wonder if we could use the existing RequestContext object? Looks like the only thing missing is the buffer supplier.

dajac replied:
That works except for the buffer supplier. I think that we can pass the buffer supplier as a regular argument instead of passing it via the context.

jolshan commented on core/src/main/scala/kafka/server/BrokerServer.scala:
```diff
@@ -22,9 +22,8 @@ import java.util
 import java.util.concurrent.atomic.AtomicBoolean
 import java.util.concurrent.locks.ReentrantLock
 import java.util.concurrent.{CompletableFuture, ExecutionException, TimeUnit, TimeoutException}
-
```
nit: did we mean to remove this line? There are a few other files that do this

dajac replied:
Nope. Reverted.

jolshan commented on clients/src/main/java/org/apache/kafka/common/requests/JoinGroupResponse.java:
```diff
@@ -28,9 +28,15 @@ public class JoinGroupResponse extends AbstractResponse {
+
+        // All versions prior to version 7 do not support nullable
+        // string for the protocol type. Empty string should be used.
+        if (version < 7 && data.protocolType() == null) {
```
I see that this was made nullable in version 7, but we are now on version 9. Did we just not check this before?

jolshan replied:
Oh I see it used to be done here in KafkaApis:
```
        val protocolName = if (request.context.apiVersion() >= 7)
          joinResult.protocolName.orNull
        else
          joinResult.protocolName.getOrElse(GroupCoordinator.NoProtocol)
```

dajac replied:
Right it was there.

jolshan commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:
```diff
@@ -0,0 +1,138 @@
+    val data = new JoinGroupRequestData()
+      .setGroupId("group")
+      .setMemberId("member")
+      .setProtocolType("consumer")
```
Did we want to add a test for the null protocol?

jolshan commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:
```diff
@@ -0,0 +1,138 @@
+
+  @ParameterizedTest
+  @ApiKeyVersionsSource(apiKey = ApiKeys.JOIN_GROUP)
+  def testJoinGroup(version: Short): Unit = {
```
Can you clarify a bit on how this test is different from the kafka apis ones?

jeffkbkim replied:
maybe a more descriptive test name would be useful here

dajac replied:
This test tests the logic of `joinGroup` method in the `GroupCoordinatorAdapter`. Given an input, it verifies the output. The tests in KafkaApisTests do not do this as they only verity that the adaptor get the expected input.

The name seems pretty clear to me as the test tests the `joinGroup` method.

jeffkbkim replied:
that makes sense. thanks!

jolshan replied:
Maybe I'm just not as well versed at reading mocked code, but that description above really helped a lot. Would it be possible to add comments in the test explaining this?

jolshan commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2528,166 @@ class KafkaApisTest {
-      testJoinGroupProtocolType(version.asInstanceOf[Short])
-    }
-  }
+  def testHandleJoinGroupRequestFutureFailed(): Unit = {
```
Is this also a new test?

dajac replied:
That's right.

jolshan reviewed (commented):
seems reasonable so far. This is just an initial pass. I will take another pass soon.

jeffkbkim commented on core/src/main/scala/kafka/coordinator/group/GroupCoordinatorAdapter.scala:
```diff
@@ -0,0 +1,81 @@
+import java.util.concurrent.CompletableFuture
+import scala.jdk.CollectionConverters._
+
+class GroupCoordinatorAdapter(
```
can we add a comment that this is a wrapper around the existing group coordinator?

jeffkbkim commented on core/src/test/scala/unit/kafka/coordinator/group/GroupCoordinatorAdapterTest.scala:
```diff
@@ -0,0 +1,138 @@
+      ("second", "second")
+    ), capturedProtocols.getValue.map { case (name, metadata) =>
+      (name, new String(metadata))
+    }.toList)
```
do we need toList?

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1646,73 +1649,46 @@ class KafkaApis(val requestChannel: RequestChannel,
     }
   }
 
+  private def makeGroupCoordinatorRequestContextFrom(
```
the naming seems slightly off. how's createGroupCoordinatorRequestContext?

dajac replied:
why do you think so? is it because of the usage of `make`? the name looks quite reasonable to me.

jeffkbkim replied:
the `From` part i guess. i don't think i've seen a method name like it, though it does make sense

dajac replied:
I can remove the `From` part. I am not too opinionated on this.

jeffkbkim reviewed (commented):
thanks for the PR, left some comments.

is the idea to add all existing group coordinator apis to GroupCoordinatorAdapter, then introduce the new GroupCoordinator which implements GroupCoordinator interface with new design, then remove GroupCoordinatorAdapter and point KafkaApis to the new GroupCoordinator?

dajac commented:
@jeffkbkim @jolshan Thanks for your comments. I have addressed your feedback. Note that I have extracted the changes about the JoinGroupResponse version handling in a separate PR: https://github.com/apache/kafka/pull/12864. It is less risky this way.

dajac committed:
address comments

dajac commented:
Rebased the PR. Ready for second round.

dajac committed:
Handle unexpected exception

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -161,6 +166,12 @@ class KafkaApis(val requestChannel: RequestChannel,
    */
   override def handle(request: RequestChannel.Request, requestLocal: RequestLocal): Unit = {
+    def handleError(e: Throwable): Unit = {
+      trace(s"Unexpected error handling request ${request.requestDesc(true)} " +
```
it looks like this message was an error level log and this change affects all other apis. what's the reason for changing it to trace?

dajac replied:
it is a mistake. reverted to error.

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1647,69 +1656,51 @@ class KafkaApis(val requestChannel: RequestChannel,
+  def handleJoinGroupRequest(
+    request: RequestChannel.Request,
+    requestLocal: RequestLocal
+  ): CompletableFuture[Unit] = {
```
some questions for my understanding:
1. it looks like we handle exceptions thrown by the new joinGroup in `newGroupCoordinator.joinGroup(ctx, joinGroupRequest.data).handle[Unit]`. are we returning a future here to handle exceptions thrown during `sendResponse()`? it looks to me that would be handled by `case e: Throwable => handleError(e)` or am i missing something?
3. would `GroupCoordinatorAdapter.joinGroup()` work without returning a CompletableFuture object? 
4. are we returning CompletableFuture in GroupCoordinatorAdapter to prepare it for the new group coordinator since it will use multiple threads to handle a single join/sync group request?

dajac replied:
1. Yeah, that's right. I return a future to catch any other errors. It could be from `sendResponse` but it could also be other issues.
2. No, it would not because the underlying API uses a callback.
3. All APIs will be async for this reason but not that most of them are already async today. The difference is that we use a future here instead of a callback.

jolshan replied:
Sorry If i missed it, but what was the reasoning of using futures instead of callbacks?

jeffkbkim replied:
i am also curious, that is one of my confusions :sweat_smile: 

> All APIs will be async for this reason but not that most of them are already async today

can you elaborate on the reason and clarify whether most of the APIs are async or not? i was under the impression that a single API request -> response is handled synchronously by a request handler thread.

dajac replied:
A future is a better abstraction than a callback in my opinion but that is subjective. We could keep using callbacks as well. The end result in the same. The new Controller interface uses Futures as well so standardizing makes sense here.

Regarding the current APIs, join group and sync group are definitely relying on callbacks. For the others, I don’t remember from the top of my head. In the new group coordinator, all of them will be executed in a different thread so they must be async.

jolshan replied:
I don't have a strong opinion on callback vs future (maybe a slight bias for me towards callbacks because I'm more familiar, but not a good reason to change the design) but I do think it makes sense to be consistent. If we plan on using futures in the new controller interface then we can go ahead with using futures here too.

jeffkbkim commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2530,208 @@ class KafkaApisTest {
+      ArgumentMatchers.eq(joinGroupRequest)
+    )).thenReturn(future)
+
+    val response = new AtomicReference[JoinGroupResponse]()
```
can you help me understand the reason for using an atomic reference?

dajac replied:
i just needed a container to store the response. i could have used a var initialized to null as well, i suppose.

jeffkbkim commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2530,208 @@ class KafkaApisTest {
+      response.set(resp)
+    }
+
+    createKafkaApis().handle(
+      requestChannelRequest,
+      RequestLocal.NoCaching
+    )
```
i'm surely missing something here - shouldn't the test be blocked here? i can't seem to find the other thread running

dajac replied:
it does not. keep in mind that we use a future internally so the method is executed and returns immediately. then, the future is completed later on by `future.completeExceptionally(Errors.NOT_COORDINATOR.exception)`.

jeffkbkim replied:
gotcha. i think i was missing this from the CompletableFuture java docs

> Actions supplied for dependent completions of non-async methods may be performed by the thread that completes the current CompletableFuture, or by any other caller of a completion method.

so the thread that completes the future will attempt to finish all (non-async) dependent chains

jeffkbkim reviewed (commented):
thanks! left some comments.

dajac committed:
address feedback

dajac commented:
@jeffkbkim Thanks for the review. I have addressed your comments.

jolshan commented on core/src/main/scala/kafka/server/KafkaServer.scala:
```diff
@@ -21,11 +21,10 @@ import java.io.{File, IOException}
 import java.net.{InetAddress, SocketTimeoutException}
 import java.util.concurrent._
 import java.util.concurrent.atomic.{AtomicBoolean, AtomicInteger}
-
```
sorry nit here again 😅

dajac replied:
I will fix this on Monday. It seems that my IDE is doing this automatically somehow.

jolshan commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1647,69 +1656,51 @@ class KafkaApis(val requestChannel: RequestChannel,
+        .format(response, request.header.correlationId, request.header.clientId))
+      requestHelper.sendResponseMaybeThrottle(request, requestThrottleMs => {
+        response.maybeSetThrottleTimeMs(requestThrottleMs)
+        response
```
If this returns unit, is there a reason we say response here?

dajac replied:
hmm… sendResponseMaybeThrottle takes a function which returns a response.

jolshan replied:
Sorry I misread the method here. 🤦‍♀️

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -1647,69 +1656,51 @@ class KafkaApis(val requestChannel: RequestChannel,
       // the group.instance.id field, so static members could accidentally become "dynamic", which leads to wrong states.
-      sendResponseCallback(JoinGroupResult(JoinGroupRequest.UNKNOWN_MEMBER_ID, Errors.UNSUPPORTED_VERSION))
+      sendResponse(joinGroupRequest.getErrorResponse(Errors.UNSUPPORTED_VERSION.exception))
+      CompletableFuture.completedFuture[Unit](())
```
`newGroupCoordinator.joinGroup()` captures exceptions inside the future returned to `handleJoinGroupRequest()` which is tested by the unit test. if we return `CompletableFuture.completedFuture()` here and in line 1694, would we be capturing the exceptions thrown during sendResponse()? not sure if that would happen, but thought it was worth mentioning since we seem to be simulating that in the unit test.

i guess i'm wondering whether we are handling these cases differently and if so, why

dajac replied:
All handlers are already surrounded by a try..catch that catches all the exceptions raised. We just need to ensure that exceptions raised in futures’ callbacks are also caught.

In this particular case, any exceptions raised by sendResponse would be caught by that try..catch.

jeffkbkim replied:
got it. thanks for the clarification

dajac committed:
fix imports

jolshan commented:
@dajac This is looking close to ready. I just had a few points.
1. I left a comment about leaving more comments on tests, not a big deal either way, but just wanted to remind.
2. I see a lot of mirror maker tests failing that I couldn't find on other branches. It is likely unrelated, but wanted to confirm.

jeffkbkim commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -184,6 +195,7 @@ class KafkaApis(val requestChannel: RequestChannel,
         case ApiKeys.OFFSET_FETCH => handleOffsetFetchRequest(request)
         case ApiKeys.FIND_COORDINATOR => handleFindCoordinatorRequest(request)
         case ApiKeys.JOIN_GROUP => handleJoinGroupRequest(request, requestLocal)
+          .exceptionally(handleError)
```
nit: i personally think it's neater to merge this with the line above, since all the other cases are single-lined

jeffkbkim reviewed (approved):
LGTM, left a minor comment. the mirror maker test failures only exist in the last commit, so i think we're good.

showuon commented on core/src/test/scala/unit/kafka/server/KafkaApisTest.scala:
```diff
@@ -2524,196 +2528,207 @@ class KafkaApisTest {
+    )
+
+    future.completeExceptionally(Errors.NOT_COORDINATOR.exception)
+    assertEquals(Errors.UNKNOWN_SERVER_ERROR, response.error)
```
nit: Could we add a comment here to mention we expected to catch the exception in `sendResponse`, and handleError will handle it? Otherwise, it is confusing why don't we get NOT_COORDINATOR error in response.

showuon commented on group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupCoordinatorRequestContext.java:
```diff
@@ -0,0 +1,74 @@
+import java.net.InetAddress;
+import java.util.Objects;
+
+public class GroupCoordinatorRequestContext {
```
nit: Should we add `toString` for debugging?

showuon reviewed (approved):
Overall LGTM! I understand why you introduce `GroupCoordinatorAdapter`, to focus on the KafkaApi changes, but I don't think that would have less change. However, I don't have strong preference for it. Let's keep on it! :)

And thanks  @jeffkbkim @jolshan for the review.

dajac committed:
address comments

dajac commented:
> I left a comment about leaving more comments on tests, not a big deal either way, but just wanted to remind.

@jolshan Thanks. I thought about this and I feel like the test is self explanatory.

dajac commented:
@jeffkbkim @jolshan @showuon Thanks for your comments. I have addressed them.

jolshan reviewed (approved):
lgtm -- thanks David!

dajac committed:
address Jason's comments

dajac commented:
I have addressed Jason's comments.

dajac committed:
fix checkstyle

hachikuji reviewed (approved):
Thanks, LGTM
